src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs

```diff
@@ -41,58 +45,106 @@ private static FileStreamStrategy ChooseStrategyCore(SafeFileHandle handle, File
-        internal static SafeFileHandle OpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options)
-            => CreateFileOpenHandle(path, mode, access, share, options);
+        internal static SafeFileHandle OpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long allocationSize)
+            => CreateFileOpenHandle(path, mode, access, share, options, allocationSize);
 
-        private static unsafe SafeFileHandle CreateFileOpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options)
+        private static unsafe SafeFileHandle CreateFileOpenHandle(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long allocationSize)
         {
-            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = GetSecAttrs(share);
+            using (DisableMediaInsertionPrompt.Create())
+            {
+                Debug.Assert(path != null);
+
+                uint ntStatus = 0;
+                if (allocationSize > 0 && (access & FileAccess.Write) != 0 && mode != FileMode.Open && mode != FileMode.Append)
+                {
+                    string prefixedAbsolutePath = PathInternal.IsExtended(path) ? path : @"\??\" + Path.GetFullPath(path); // we might consider getting rid of this managed allocation
+                    (ntStatus, IntPtr fileHandle) = Interop.NtDll.CreateFile(prefixedAbsolutePath, mode, access, share, options, allocationSize);
+                    if (ntStatus == 0)
+                    {
+                        return ValidateFileHandle(new SafeFileHandle(fileHandle, ownsHandle: true), path, (options & FileOptions.Asynchronous) != 0);
+                    }
+                    else if (ntStatus == NT_ERROR_STATUS_DISK_FULL)
+                    {
+                        throw new IOException(SR.Format(SR.IO_DiskFull_Path_AllocationSize, path, allocationSize));
+                    }
+                    else if (ntStatus == NT_ERROR_STATUS_FILE_TOO_LARGE)
+                    {
+                        throw new IOException(SR.Format(SR.IO_FileTooLarge_Path_AllocationSize, path, allocationSize));
+                    }
 
-            int fAccess =
-                ((access & FileAccess.Read) == FileAccess.Read ? Interop.Kernel32.GenericOperations.GENERIC_READ : 0) |
-                ((access & FileAccess.Write) == FileAccess.Write ? Interop.Kernel32.GenericOperations.GENERIC_WRITE : 0);
+                    // NtCreateFile has failed for some other reason than a full disk or too large file.
+                    // Instead of implementing the mapping for every NS Status value (there are plenty of them: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55)
+                    // or using RtlNtStatusToDosError & GetExceptionForWin32Error
+                    // the code falls back to CreateFileW that just throws the right exception.
```

Comment:
I assume you already tried to find an "NS Status to Win32 exception" converter in our codebase and didn't find any?

Reply:
I've found some, but I am just afraid that this could change the exception that was thrown so far when `allocationSize` was not specified

Reply:
This fallback can lead to the allocation size being intermittently ignored (e.g. if the first call fails on file exist, but the fallback succeeds because of the file was deleted in the meantime).

The mapping from NT status codes to Win32 codes is very simple. CreateFile has an special case for `STATUS_OBJECT_NAME_COLLISION` and `STATUS_FILE_IS_A_DIRECTORY` that you can just replicate here.

I would more worried about other special handling that CreateFile does on top of NtCreateFile. Have you checked that none of that is needed?

Reply:
> if the first call fails on file exist, but the fallback succeeds because of the file was deleted in the meantime

Great catch. This has forced me to re-think the approach and change it in a way that if `NtCreateFile` fails, we map the error code using `RtlNtStatusToDosError` and then throw an exception.

> Have you checked that none of that is needed?

I did and this is now covered by unit tests:

https://github.com/dotnet/runtime/blob/c2f995efd5a26b2501e71df173894da7d0a707ae/src/libraries/System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs_buffer_fo_as.cs#L8

(`FileStream_ctor_str_fm_fa_fs_buffer_fo` derives from `FileStream_ctor_str_fm_fa_fs_buffer`, which derives from `FileStream_ctor_str_fm_fa_fs` and so on).

These tests have helped me to detect many non-obvious dependencies. In the mapping code everything that has a comment is an argument that required special handling:

https://github.com/dotnet/runtime/blob/c2f995efd5a26b2501e71df173894da7d0a707ae/src/libraries/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs#L73-L176